sentry/src/main/java/io/sentry/clientreport/ClientReportRecorder.java

```diff
@@ -0,0 +1,171 @@
+
+import io.sentry.DateUtils;
+import io.sentry.SentryEnvelope;
+import io.sentry.SentryEnvelopeItem;
+import io.sentry.SentryItemType;
+import io.sentry.SentryLevel;
+import io.sentry.SentryOptions;
+import io.sentry.transport.DataCategory;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public final class ClientReportRecorder implements IClientReportRecorder {
+
+  private static volatile @Nullable ClientReportRecorder SHARED_INSTANCE = null;
+
+  public static @NotNull ClientReportRecorder getInstance() {
+    if (SHARED_INSTANCE == null) {
+      synchronized (ClientReportRecorder.class) {
+        if (SHARED_INSTANCE == null) {
+          SHARED_INSTANCE = new ClientReportRecorder();
+        }
+      }
+    }
+
+    return SHARED_INSTANCE;
+  }
+
+  private final ClientReportStorage storage;
+
+  private ClientReportRecorder() {
+    this.storage = new AtomicClientReportStorage();
+  }
+
+  @Override
+  public @NotNull SentryEnvelope attachReportToEnvelope(
+      @NotNull SentryEnvelope envelope, @NotNull SentryOptions options) {
+    @Nullable ClientReport clientReport = resetCountsAndGenerateClientReport();
+    if (clientReport == null) {
+      return envelope;
+    }
+
+    try {
+      options.getLogger().log(SentryLevel.DEBUG, "Attaching client report to envelope.");
+
+      final List<SentryEnvelopeItem> items = new ArrayList<>();
+
+      for (final SentryEnvelopeItem item : envelope.getItems()) {
+        items.add(item);
+      }
+
+      items.add(SentryEnvelopeItem.fromClientReport(options.getSerializer(), clientReport));
+
+      return new SentryEnvelope(envelope.getHeader(), items);
+    } catch (Throwable e) {
+      options.getLogger().log(SentryLevel.ERROR, e, "Unable to attach client report to envelope.");
+      return envelope;
+    }
+  }
+
+  @Override
+  public void recordLostEnvelope(
+      @NotNull DiscardReason reason,
+      @Nullable SentryEnvelope envelope,
+      @NotNull SentryOptions options) {
+    if (envelope == null) {
+      return;
+    }
+
+    try {
+      for (final SentryEnvelopeItem item : envelope.getItems()) {
+        recordLostEnvelopeItem(reason, item, options);
+      }
+    } catch (Throwable e) {
+      options.getLogger().log(SentryLevel.ERROR, e, "Unable to record lost envelope.");
+    }
+  }
+
+  @Override
+  public void recordLostEnvelopeItem(
+      @NotNull DiscardReason reason,
+      @Nullable SentryEnvelopeItem envelopeItem,
+      @NotNull SentryOptions options) {
+    if (envelopeItem == null) {
+      return;
+    }
+
+    try {
+      final @NotNull SentryItemType itemType = envelopeItem.getHeader().getType();
+      if (SentryItemType.ClientReport.equals(itemType)) {
+        try {
+          final ClientReport clientReport = envelopeItem.getClientReport(options.getSerializer());
+          restoreCountsFromClientReport(clientReport);
```

Comment:
Q: Is this something we might need on Dart/Flutter? @marandaneto 

Q: Are we ignoring/dropping the `reason` method param when we are restoring the count from a client report envelope item? It's also the behaviour that the unit test expects.

Reply:
That's likely going to be done only by Cocoa and Android since the Flutter and React Native SDK only write an event to the disk.
Maybe check if this is needed for `HttpTransport` in the pure Dart client.

Reply:
> Are we ignoring/dropping the reason method param when we are restoring the count from a client report envelope item? It's also the behaviour that the unit test expects.
We're not really tracking the lost Client Report in that case. If we were, we'd use the reason param for that but since it just restores counts we want to keep the original reason imho.

Reply:
The API is just a bit ambiguous, IHMO. When the item happens to be a client report, the `reason` just gets ignored. Not necessarily something the caller knows about, as it's a hidden implementation / side effect.

Reply:
We don't restore that in Cocoa. We just ignore dropping envelope items with client reports. Again, client reports is best effort, and it's fine to lose some data. We don't have to cover all edge cases.